### ISSUE TITLE
Compare two fingerprints.

### ISSUE THREAD

**Issue by avatar29A** [beetbox/pyacoustid#18]
Another question. Maybe offtopic, but still. I have two identical tracks from different sources. Physically, the tracks are different (possibly encoding compression). I get their fingerprints. They are of course different. Is it possible to get some distance between fingerprints to compare their similarities. If so, in which direction to look? Thank you.

**Comment by naggie** [comment]
```
def compare_b64_fingerprint(a,b):
	'returns inverse bit error rate in percent'
	a = b64decode(a)
	b = b64decode(b)

	matching_bits  = 0
	different_bits = 0

	if len(b) != len(a):
		# trim to shortest
		shortest = min(len(a),len(b))
		longest  = max(len(a),len(b))

		a = a[:shortest]
		b = b[:shortest]

		# extra bits count as different
		different_bits = 8*(longest-shortest)

	for i in range( len(a) ):
		byte_a = ord( a[i:i+1] )
		byte_b = ord( b[i:i+1] )

		for x in range(8):
			bit_a = byte_a >> x & 1
			bit_b = byte_b >> x & 1

			if (bit_a == bit_b):
				matching_bits  +=1
			else:
				different_bits +=1

	total_bits = different_bits + matching_bits

	return float(matching_bits) / float(total_bits)

```

This is what I used in a python library that simply used the executable. I put it here in case anyone wants it.

**Comment by lalinsky** [comment]
I don't think that code does what you expect. What are the inputs to the function? For comparing with fingerprints, you need an list of 32-bit integers. Here you are taking some base64-encoded string, decode it and then work with it one byte at a time.

Also, this kind of bit manipulations in Python are very slow. At the very least you should use an 8-bit popcount lookup table, e.g. https://gist.github.com/lalinsky/1132166

**Comment by naggie** [comment]
Base64 strings, as I said it was with an executable, not the library. Not optimal but it worked bitwise. 

Thanks for the link, I'll keep it in mind.

**Comment by lalinsky** [comment]
Which binary, fpcalc? The base64-encoded strings from there are compressed and you can't directly compare them. You need to use `fpcalc -raw` if you don't want to use the library to uncompress them.

**Comment by lalinsky** [comment]
Oh, so these are not chromaprint fingerprints?

**Comment by naggie** [comment]
They are base64 encoded chromaprints. Produced by an `fpcalc` modified to produce base64 encoded output instead of a series of ints....

The encoding is irrelevant, the method used to calculate BER is relevant here.

**Comment by lalinsky** [comment]
Well, I clearly can't convince you, but the code you are using is not comparing the audio fingerprints, only their compressed versions. If it works for you, it's just by luck. What you are doing is like taking two similar text files, compressing them with zip and then comparing the zip files. Even though the original text files were similar, the zip files are completely different.

**Comment by naggie** [comment]
What? No, I have 2 regular fingerprints which are base64 encoded **in a completely separate custom executable**. I then run the executable **in a python subprocess,** and run this on the output. Why do you think I'm using zip files?

**Comment by naggie** [comment]
The chromaprint is generated by this executable. What's not to get?

**Comment by naggie** [comment]
I'm not comparing the raw audio files myself. The custom executable is generating a fingerprint for each song.

**Comment by naggie** [comment]
@lalinsky understand?

**Comment by lalinsky** [comment]
I'm not sure why you posted the code example here, because if it's comparing output of your own program, it's not of much use who use standard acoustid/chromaprint tools.

**Comment by naggie** [comment]
The algorithm for the bit error rate calculation is relevant here, it can be applied to any program that uses chromaprints -- if you ignore the bit that converts from base64.

**Comment by shidarin** [comment]
This would be really useful in pyacoustid.

**Comment by salamanders** [comment]
This result came up in my searching, I thought I'd add where I ended up in case it helps others.

1. Cache a list of ints for each mp3 using `fpcalc -raw -signed`
2. M^2 comparisons using kotlin:

```kotlin
fun chromaDistance(cl0: List<Int>, cl1: List<Int>): Double = cl0.zip(cl1).map { (c0, c1) ->
    (c0 xor c1).countOneBits()
}.also {
    require(it.isNotEmpty())
}.average()
```

And it felt like matches < 10.0 started to find duplicates.

**Comment by geri777** [comment]
I found this Python library which utilizes fpcalc and does a great job:
https://medium.com/@shivama205/audio-signals-comparison-23e431ed2207

**Comment by salamanders** [comment]
> I found this Python library which utilizes fpcalc and does a great job:
> https://medium.com/@shivama205/audio-signals-comparison-23e431ed2207

AH!  Offsets.  I forgot offsets.  Dangit.  Thank you for the link!

**Comment by bindestriche** [comment]
I took a stab at it. Decompressed the fingerprints, as @lalinsky pointed out. I tried to port [acoustid_compare.c](https://github.com/acoustid/pg_acoustid/blob/main/acoustid_compare.c) . As such, I lay no claim to intellectual property to the code below, if it's correct feel free to include it in this lib.


```py
import numpy as np
import acoustid
ACOUSTID_MAX_BIT_ERROR = 2
ACOUSTID_MAX_ALIGN_OFFSET = 120

def popcount(x):
    return bin(x).count('1')

def match_fingerprints(a, b):
    asize = len(a)
    bsize = len(b)
    numcounts = asize + bsize + 1
    counts = np.zeros(numcounts, dtype=int)

    for i in range(asize):
        jbegin = max(0, i - ACOUSTID_MAX_ALIGN_OFFSET)
        jend = min(bsize, i + ACOUSTID_MAX_ALIGN_OFFSET)
        for j in range(jbegin, jend):
            biterror = popcount(a[i] ^ b[j])
            if biterror <= ACOUSTID_MAX_BIT_ERROR:
                offset = i - j + bsize
                counts[offset] += 1

    topcount = counts.max()
    return topcount / min(asize, bsize)

def compare_fingerprints(a, b):
    import base64
    # pad
    from acoustid import chromaprint
    a=chromaprint.decode_fingerprint(a)[0]
    b=chromaprint.decode_fingerprint(b)[0]
    a = np.array(a, dtype=np.int32)
    b = np.array(b, dtype=np.int32)
    return match_fingerprints(a, b)

#use
_, a = acoustid.fingerprint_file(file_path1)
_, b = acoustid.fingerprint_file(file_path2)
compare_fingerprints(a,b)
```

**Comment by albertopasqualetto** [comment]
It is now implemented in the library, so I think this issue can be closed.
